metals/src/main/scala/scala/meta/internal/metals/callHierarchy/CallHierarchyProvider.scala

```diff
@@ -0,0 +1,619 @@
+              )
+            )
+          }
+        )
+        .flatten
+        .toList
+    )
+
+  /**
+   * Resolve incoming calls for a given call hierarchy item.
+   */
+  def incomingCalls(
+      params: CallHierarchyIncomingCallsParams,
+      token: CancelToken,
+  )(implicit ec: ExecutionContext): Future[List[CallHierarchyIncomingCall]] = {
+    val info = params.getItem.getData
+      .asInstanceOf[JsonElement]
+      .as[CallHierarchyItemInfo]
+      .get
+
+    for {
+      paths <-
+        if (!info.isLocal)
+          references
+            .pathsMightContainSymbol(
+              params.getItem.getUri.toAbsolutePath,
+              info.symbols.toSet,
+            )
+        else Future.successful(Set.empty)
+      localPath = if (info.searchLocal) Set(workspace) else Set.empty
+      result <- Future.sequence(
+        (paths ++ localPath).toList
+          .map(source => {
+            semanticdbs.textDocument(source).documentIncludingStale match {
+              case Some(doc) =>
+                val results = trees
+                  .get(source)
+                  .map(root =>
+                    findIncomingCalls(
+                      source,
+                      doc,
+                      root,
+                      info,
+                    ) ++ findIncomingCallsSynthetics(
+                      source,
+                      doc,
+                      info,
+                    )
+                  )
+                  .getOrElse(Nil)
+
+                Future
+                  .sequence(results.map {
+                    case FindIncomingCallsResult(
+                          occurence @ SymbolOccurrence(_, symbol, _),
+                          range,
+                          ranges,
+                        ) if !containsDuplicates(info.visited) =>
+                      buildCallHierarchyItem(
+                        source,
+                        doc,
+                        occurence,
+                        range,
+                        info.visited :+ symbol,
+                        token,
+                      )
+                        .mapOptionInside(chi =>
+                          new CallHierarchyIncomingCall(
+                            chi,
+                            ranges.asJava,
+                          )
+                        )
+                    case _ => Future.successful(None)
+                  })
+                  .map(_.flatten)
+              case None =>
+                Future.successful(Nil)
+            }
+          })
+      )
+    } yield (result.flatten)
+  }
+
+  private def findDefinitionOccurence(
+      source: AbsolutePath,
+      symbol: String,
+  )(implicit
+      ec: ExecutionContext
+  ): Future[Option[(SymbolOccurrence, AbsolutePath, TextDocument)]] =
+    (if (!symbol.isLocal)
+       references
+         .pathsMightContainSymbol(
+           source,
+           Set(symbol),
+         )
+     else Future.successful(Set(source)))
```

Comment:
Just a nit, my preference is typically to do the operations you need, and then at the end if it needs to be returned in a `Future`, then do this at the end.